src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,204 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> newNode = new Node(value, tail, null);
+        if (isEmpty()) {
+            head = newNode;
+        } else {
+            tail.setNext(newNode);
+        }
+        tail = newNode;
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) { //add on tail
+            add(value);
+            return;
+        }
+
+        Node<T> nodeBefore = null;
+        Node<T> nodeAfter = head;
+
+        if (index == 0) { //add on head
+            Node<T> newNode = new Node(value, nodeBefore, nodeAfter);
+            nodeAfter.setPrev(newNode);
+            head = newNode;
+            size++;
+            return;
+        }
+
+        if (isIndexInvalid(index)) {
+            throw new IndexOutOfBoundsException();
+        }
 
+        for (int i = 0; i < index; i++) {
+            nodeAfter = nodeAfter.getNext();
+        }
+        nodeBefore = nodeAfter.getPrev();
+        Node<T> newNode = new Node(value, nodeBefore, nodeAfter);
+        nodeBefore.setNext(newNode);
+        nodeAfter.setPrev(newNode);
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexInvalid(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        Node<T> result = head;
+        for (int i = 0; i < index; i++) {
+            result = result.getNext();
+        }
```

Comment:
can we move this logic to the separate private method? and use it in the methods: remove(), get(), set(), addByIndex()